internal/corazawaf/transaction.go

```diff
@@ -748,8 +749,75 @@ func (tx *Transaction) ProcessRequestHeaders() *types.Interruption {
 	return tx.interruption
 }
 
-func (tx *Transaction) RequestBodyWriter() io.Writer {
-	return tx.RequestBodyBuffer
+type Lenger interface {
+	Len() int
+}
+
+func (tx *Transaction) WriteRequestBody(b []byte) (*types.Interruption, int, error) {
+	if tx.RuleEngine == types.RuleEngineOff {
+		return nil, 0, nil
+	}
+
+	writingBytes := int64(len(b))
+	if tx.requestBodyBuffer.length+writingBytes >= tx.RequestBodyLimit {
+		tx.variables.inboundErrorData.Set("1")
+		if tx.WAF.RequestBodyLimitAction == types.RequestBodyLimitActionReject {
+			// We interrupt this transaction in case RequestBodyLimitAction is Reject
+			tx.interruption = &types.Interruption{
+				Status: 403,
+				Action: "deny",
+			}
+			return tx.interruption, 0, nil
+		}
+
+		if tx.WAF.RequestBodyLimitAction == types.RequestBodyLimitActionProcessPartial {
+			writingBytes = tx.RequestBodyLimit - tx.requestBodyBuffer.length
+		}
+	}
+
+	w, err := tx.requestBodyBuffer.Write(b[:writingBytes])
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return nil, int(w), nil
+}
+
+// ProcessRequestBodySize sets the maximum number of bytes that can be read from the
+// request body.
+func (tx *Transaction) WriteRequestBodyFrom(r io.Reader) (*types.Interruption, int, error) {
```

Comment:
`ReadRequestBodyFrom` - same as `io.Write/ReadFrom`

Reply:
In the end I find this method a bit problematic. If we can't figure out the body length on beforehand then we will never know if triggering the interruption for body limit as we will read up to a limit and in order to know if the remaining bytes are in the reader or not we need to read the reader and by that modifying it.

For example, if the body size is 100 and the limit is 20, this method will read up to 20 and then leave 80 remaining in the body reader but we won't know unless we attempt to read but if we attempt to read (let's say 1 byte) we will modify the body reader to 79.